### DISCUSSION TITLE
Local HTTPS server for development

### DISCUSSION THREAD

**Discussion by justin-dunham** [vercel/next.js#27040] [Ideas]
### Describe the feature you'd like to request

What?
There currently is not a built in way to use https while in development mode.

Why?
This is required for testing some browser APIs especially on local devices. 

How?
Other examples of this optional feature include webpack dev server
https://webpack.js.org/configuration/dev-server/#devserverhttps

### Describe the solution you'd like

Would love to see even a self-signed cert with local https. This would allow local network devices to connect to the development server for testing purposes.  

### Describe alternatives you've considered

There are some browser flags or options to disable security. This seems like a risk to devices if they are used for any other purpose than local testing.

**Comment by theetrain** [comment]
While there is no built-in method for HTTPS, you can enable it yourself using a custom server. You can view the official NextJS [guide for custom servers](https://nextjs.org/docs/advanced-features/custom-server) to choose the server style you prefer.

Then, you'll need to generate your own local certificates and apply them using Node's [https.createServer](https://nodejs.org/api/https.html#https_https_createserver_options_requestlistener) method.

Here's an example I managed to get working for myself using Express: https://gist.github.com/theetrain/a74ce0656a0ca967ac20ee2a136226ae

It involves these considerations:

1. Create certificates using `mkcert` and save them under a gitignored `.cert` directory (you can follow the guide here: https://dev.to/nakib/using-https-on-next-js-local-development-server-bcd, but note my gist above differs slightly from the example in this article)
2. Add `APP_DOMAIN=[your development IP or domain]` and `HTTPS=true` in your `.env.development` file
3. Run the server using `APP_ENV=development node server.js`

At the time of this writing, I am using:

express@4.17.1
next@10.2.3
npm@6.14.13
Node.js v14.17.0

**Reply by kuhlaid** [reply]
Is there a solution that does not involve a custom server (which breaks built-in NextJS API routes)?

**Comment by stigkj** [comment]
You could use a proxy in front of the Next.js app that does SSL termination. For example https://github.com/cortesi/devd, i.e.
```
> devd --tls /=http://localhost:<port>
```

**Reply by stigkj** [reply]
`--tls` generates a certificate for `*devd.io` which points to localhost. You could use `mkcert` (https://github.com/FiloSottile/mkcert) to generate your own certificate for your own hostname and then use `devd --cert=<path to cert>`.
